### PR TITLE
Implement unwinding for i686 MSVC

### LINKED ISSUE #25869: MSVC: Implement 32-bit Unwinding
## Updated report

* [x] 64-bit unwinding
* [ ] 32-bit unwinding (see [comment below](https://github.com/rust-lang/rust/issues/25869#issuecomment-141811103))

## Original Report

Right now the MSVC target turns all calls to `panic!` into a process-wide abort as unwinding is not implemented. 

* [x] Update LLVM (https://github.com/rust-lang/rust/pull/26025)
* [ ] https://llvm.org/bugs/show_bug.cgi?id=23884

WIP branch: ~~https://github.com/alexcrichton/rust/tree/msvc-llvm-update~~ (now merged)

### PR DESCRIPTION
**Pull request by alexcrichton**
These commits perform a few high-level changes with the goal of enabling i686 MSVC unwinding:
- LLVM is upgraded to pick up the new exception handling instructions and intrinsics for MSVC. This puts us somewhere along the 3.8 branch, but we should still be compatible with LLVM 3.7 for non-MSVC targets.
- All unwinding for MSVC targets (both 32 and 64-bit) are implemented in terms of this new LLVM support. I would like to also extend this to Windows GNU targets to drop the runtime dependencies we have on MinGW, but I'd like to land this first.
- Some tests were fixed up for i686 MSVC here and there where necessary. The full test suite should be passing now for that target.

In terms of landing this I plan to have this go through first, then verify that i686 MSVC works, then I'll enable `make check` on the bots for that target instead of just `make` as-is today.

Closes #25869

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @rust-lang/compiler, there's some pretty significant trans changes here!
cc @vadimcn, lots of windows changes
cc @brson, build system changes! (and llvm changes)

**Comment by alexcrichton** [comment]
Oh and I think it's important to wait for travis before r+'ing this b/c that make sure that we're still compatible with LLVM 3.7 (which I believe is important for now)

**Comment by sanxiyn** [comment]
Travis failed compiling RustWrapper.cpp on LLVM_VERSION_MINOR < 8 paths.

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -686,5 +686,8 @@ extern "rust-intrinsic" {
     /// Rust's "try catch" construct which invokes the function pointer `f` with
     /// the data pointer `data`, returning the exception payload if an exception
     /// is thrown (aka the thread panics).
+    #[cfg(not(stage0))]
+    pub fn try(f: fn(*mut u8), data: *mut u8, local_ptr: *mut u8) -> i32;
```
Comment (nikomatsakis) [inline src/libcore/intrinsics.rs]: maybe update the comment with what the heck `local_ptr` is?

**Comment by vadimcn** [comment]
Nice work, @alexcrichton!

After the first pass this looks fine to me. 
I'll try to find some time to build this branch and play with it.  I might have more insights then.

File: src/libcore/intrinsics.rs (hunk 1 above)
Comment (alexcrichton) [inline src/libcore/intrinsics.rs]: Surely!

**Comment by brson** [comment]
Impressive.

**Comment by alexcrichton** [comment]
@nikomatsakis brought up some interesting points on IRC that I didn't realize here. It turns out that today we actually cache cleanups for landing pads, for example with code like:

```rust
pub fn foo() {
    let _a = Foo;
    bar();
    let _b = Foo;
    bar();
}
```

There are two landing pads for this function from the two calls to `bar`, and both landing pads need a cleanup of the local variable `_a`. Today we only translate the cleanup of `_a` once, however, and the landing pad of `_b` will branch to the cleanup of `_a`. This is visible in the [generated IR](https://gist.github.com/alexcrichton/88039f077021405754a9).

A problem here, however, is that with `cleanuppad` this is not possible. The same basic block (e.g. the drop of `_a`) is being shared between two `cleanuppad` instances, so it doesn't clearly belong to either one to add a funclet bundle to it. This branch currently works accidentally by basically breaking the cache. For example the IR of this PR re-translates the drop of `_a` twice.

I just remembered, however, that the [`cleanupret` instruction](http://llvm.org/docs/LangRef.html#cleanupret-instruction) can take a label to continue unwinding to. I believe we can use this to chain together cleanup pads in a way similar to how `landingpad`/GNU works. 

This'll take some work to get working, but seems critical enough to get it done before landing.

**Comment by alexcrichton** [comment]
Ok, I've now pushed an updated version which handles caching better on MSVC. The caching for GNU/landingpad exceptions should be the same as it was before, and the [IR for the snippet above](https://gist.github.com/alexcrichton/555a0292525c1131d181) in MSVC shows the caching being used as well.

Currently the caching isn't *as* aggressively used on MSVC as it is on GNU b/c on MSVC you can't jump to the middle of a set of cleanup scopes, only the front. Perhaps that only matters for corner cases though?

**Comment by nikomatsakis** [comment]
@alexcrichton 

> Currently the caching isn't as aggressively used on MSVC as it is on GNU b/c on MSVC you can't jump to the middle of a set of cleanup scopes, only the front. Perhaps that only matters for corner cases though?

So I skimmed the code and sort of grok it. But I don't quite see where this comes in yet. What makes one version more aggressive than the other?

**Comment by alexcrichton** [comment]
@nikomatsakis, @nagisa 

Updated with comments addressed hopefully!

> What makes one version more aggressive than the other?

This has to do with [this comment](https://github.com/alexcrichton/rust/blob/9cf0e0673c41b3d6b3a619e56bbe0ed323ddc122/src/librustc_trans/trans/cleanup.rs#L862-L879). The gist of it is that if there are multiple entries in `popped_scopes` when generating landing pads, only the *last* one will be used as a "cached early exit" for MSVC. With GNU exceptions all the intermediate ones will be used as well. This is because with GNU exceptions we can jump to any other basic block and keep going, but with MSVC exceptions we can't do that because each basic block can be a member of only one "funclet", so we have to be able to statically know which basic blocks are part of which cleanup pad chain.

Does that make sense? I fear that this is a little obscure and I have perhaps not commented enough, so please don't hesitate to ask for clarification!

**Comment by nikomatsakis** [comment]
On Mon, Jan 11, 2016 at 09:54:44PM -0800, Alex Crichton wrote:
> Does that make sense? I fear that this is a little obscure and I have perhaps not commented enough, so please don't hesitate to ask for clarification!

Yes, but a quick point of clarification. Earlier you hinted that there is in fact a way to chain funclets by "throwing" to another funclet or something like that
(can't be bothered to look up the exact terms). This is being deferred for future work, right? (Sorry, PR is out of cache just now)

**Comment by alexcrichton** [comment]
Ah no worries, having funclets "throws to other funclets" was actually implemented, and it looks like this: https://github.com/alexcrichton/rust/blob/9cf0e0673c41b3d6b3a619e56bbe0ed323ddc122/src/librustc_trans/trans/cleanup.rs#L780-L803

**Comment by alexcrichton** [comment]
So I was looking at the LLVM history today and there's been quite a number of WinEH commits since the LLVM revision I pinned this to. Many of them were specifically around verification of funclets, so I decided to update LLVM and see what happened. That update is not currently part of this PR, and I don't *think* that it will be necessary, but I learned something interesting I wanted to share!

So far I ended up [just running into an LLVM bug](https://llvm.org/bugs/show_bug.cgi?id=26147) during optimizing libstd, but it brought up a [very interesting comment](https://llvm.org/bugs/show_bug.cgi?id=26147#c1). The gist of it (as I understand) is that a `call` instruction in a funclet which leads to unwinding is considered undefined behavior. Note that we have many `call` instructions in funclets (e.g. calling drop glue, and notably not `invoke`s), and destructors are definitely allowed to panic (although seriously advised against). We're safe from undefined behavior, however, because we can only ever be part of a funclet via a panic, and then a subsequent panic will abort the entire process.

What this means, I believe, is that on MSVC it will be required for memory safety to have double panics lead to aborts for now. We can fix this by having funclets use `invoke` instead of `call` and properly handling a panicking destructor in a panicking destructor, but it's largely just something to keep in mind.

**Comment by nikomatsakis** [comment]
On Thu, Jan 14, 2016 at 10:08:12PM -0800, Alex Crichton wrote:
> What this means, I believe, is that on MSVC it will be required for memory safety to have double panics lead to aborts for now. We can fix this by having funclets use `invoke` instead of `call` and properly handling a panicking destructor in a panicking destructor, but it's largely just something to keep in mind.

Interesting. I think I've come to the conclusion that I would want
dtors to be "noexcept" (i.e., if unwinding reaches the dtor, it
aborts) as well, at least for consistency. I wonder if we can make
that change.

**Comment by nagisa** [comment]
> I wonder if we can make that change.

Not sure if that’s a valid way forward. Panicking (and thus unwinding from) in destructors is valid, as long as the destructor is only called during regular cleanups and not unwind-cleanups.

We could make an RFC and a bunch of lints to forbid unwinding from destructors though, but I’m not sure it is considerable at all as pre-2.0 material.

**Comment by alexcrichton** [comment]
Ok, I've pushed the upgrade to LLVM which actually now puts us on the `release_38` branch of LLVM. This should mean it'll be pretty easy for us to track changes there as 3.8 is release in the (hopefully) near future.

I've implemented [the recommended workaround](https://llvm.org/bugs/show_bug.cgi?id=26147#c3) for the LLVM bug for now, which is to annotate all `call` instructions with `nounwind` if they happen in a cleanup pad. Note that this is indeed true at runtime because a second panic rooted in a cleanup pad is a double-panic which will abort the process.

**Comment by alexcrichton** [comment]
The specific change [is here as well](https://github.com/alexcrichton/rust/blob/llvmup/src/librustc_trans/trans/build.rs#L931-L952)

**Comment by alexcrichton** [comment]
Turns out we've got some MIR tests which are exercising unwinding functionality, but as these changes aren't ported over to MIR just yet I'm going to disable those tests on MSVC and I've opened up https://github.com/rust-lang/rust/issues/30941 to track the implementation.

**Comment by nikomatsakis** [comment]
@nagisa traditionally we have said that the semantics of panics in a dtor
are an area we reserve the right to change (eg in the language semver rfc),
but it may be that we would break too much working code to actually change
things there. This is why I said "I wonder if we can make the change...."

**Comment by nagisa** [comment]
I’ve submitted a counter-PR against alexcrichton:llvm with a basic MIR implementation of this thing.

**Comment by alexcrichton** [comment]
rebased and chatted on IRC with @nagisa about the MIR implementation, it looks like it's missing the "all call instructions must be annotated with the funclet they're in" temporarily, so I'm gonna hold off on merging just yet

**Comment by nagisa** [comment]
I discovered additional difficulties implementing this in MIR and will land MIR implementation independently once this PR lands.

I overall like the new instructions, and I find the changes proposed by this PR okay. I dislike us maintaining so much extra function-wide state in regards to the additional EH implementation, but I don’t immediately see any obvious way around it.

**Comment by shepmaster** [comment]
> Ok, I've pushed the upgrade to LLVM which actually now puts us on the release_38 branch of LLVM. This should mean it'll be pretty easy for us to track changes there as 3.8 is release in the (hopefully) near future.

I mentioned this on IRC, but if the going gets rough for the unwinding parts of this PR, could the LLVM upgrade aspect be split out? I'm attempting to track some of the [changes for llvm-avr](https://github.com/avr-rust/rust/issues/14) which is also tracking LLVM HEAD.

**Comment by alexcrichton** [comment]
I moved us further along the LLVM 3.8 release branch which includes a fix for the issue I found above, which means that there's no longer any need to tag `call` instructions in a landing pad as `nounwind` (although we can probably continue to do this anyway due to our double-panic-is-abort situation.

**Comment by nikomatsakis** [comment]
OK, so, I read through this branch again. I also talked some to @alexcrichton offline. Basically, I think we can do better with the re-use here. In particular, if you have four nested scopes, like:

```
SCOPE_A
SCOPE_B
SCOPE_C
SCOPE_D
```

then what I expect is that we should generate one LLVM "funclet" per scope, and use the `cleanupret` instruction to chain between them. What this branch actually does is to (sometimes) put multiple scopes in the funclets. What I'd prefer to see would thus look like:

```
UNWIND_A: 
    token_a = cleanuppad();
    cleanupret(token_a, caller);

UNWIND_B:
    token_b = cleanuppad();
    cleanupret(token_b, UNWIND_A);

UNWIND_C:
    token_c = cleanuppad();
    cleanupret(token_c, UNWIND_B)

UNWIND_D:
    token_d = cleanuppad
    cleanupret(token_d, UNWIND_C);
```

In terms of MIR, it seems like we should basically adopt this "funclet" model -- marking basic blocks as either normal, or tagging them as "unwind" (and carrying along the "funclet" code, rather than making a fake value to represent it). This is trivially translatable to MSVC and trivially erasable to GNU (with branches).

**Comment by nikomatsakis** [comment]
(That said, I'd be happy to leave that as a FIXME)

**Comment by alexcrichton** [comment]
@nikomatsakis ok I think I've implemented what we talked about. I added a few helper functions to `cleanup.rs` and I think it's a little cleaner how the changes were applied (and easier to follow the control flow). It's at least a little clearer to me now that the logic is basically the same and as much caching happens afterwards as it does before.

The IR [for a simple function](https://gist.github.com/alexcrichton/15c06b98e122a12d707e) is [pretty awful](https://gist.github.com/alexcrichton/82dad941d9b6710b9753) as you can see [cleanuppads are *immediately* followed by cleanupret](https://gist.github.com/alexcrichton/82dad941d9b6710b9753#file-foo-ll-L52-L53), as well as tons of branching via `cleanupret`. That being said, however, it looks like it's legal from LLVM's point of view and it also does [quite a good job optimizing it](https://gist.github.com/alexcrichton/8bc67b8215a5caad6435) into basically exactly what one might expect.

**Comment by nikomatsakis** [comment]
@alexcrichton indeed, it looks nicer now. I am pleased. Thanks for putting in the effort.

**Comment by alexcrichton** [comment]
Ugh, I forgot about this. We're going to need to upgrade the gcc/g++ on the windows bots as apparently 4.9.3 doesn't have `std::thread` in the C++ standard library which LLVM now requires. I'll probably just pick the newest toolchain (which I think is 5.3.0) from the mingw-w64 project.

cc @vadimcn, @brson, may mean some upcoming distribution issues!

**Comment by alexcrichton** [comment]
As @vadimcn pointed out on IRC https://github.com/rust-lang/rust/issues/28260 may throw a wrench in the works, but who knows! (I certainly don't...)

**Comment by alexcrichton** [comment]
Ok, I've done a few things:

1. I've created two temporary commits in our LLVM branch
    * https://github.com/rust-lang/llvm/commit/ae6a38329e0b40ed5a7bd1d31ffe0ba48ead1cc7 is a cherry-pick of https://github.com/llvm-mirror/llvm/commit/707c12c224206155ede6c172df8d264d43ddd4c1 which just isn't in the `release_38` branch at this time.
    * https://github.com/rust-lang/llvm/commit/664319f7cca6374c320adf0ebe38305bca50ac68 is my attempt to get LLVM to compile on mingw-w64 by just not compiling any part of LLVM that needs `std::thread`. It looks like all portions aren't used by Rust itself at least. Note that this is similar to https://github.com/rust-lang/llvm/commit/0be2a329e6dff0b6996ab256e0c1d39da975a519 in that it's not our first "let's make LLVM compile everywhere we need" patch.
2. As we're not using `std::thread`, I'm not upgrading gcc on the bots

@bors: r+ 7f0d1f7fdfe65c1aa85a7bd7c81f2cf2617e0e02

**Comment by alexcrichton** [comment]
er

@bors: r=nikomatsakis 7f0d1f7fdfe65c1aa85a7bd7c81f2cf2617e0e02

**Comment by vadimcn** [comment]
> As we're not using std::thread, ...

We aren't?  I thought it was needed for parallel codegen?

**Comment by alexcrichton** [comment]
@bors: r=nikomatsakis 2c58835

> We aren't? I thought it was needed for parallel codegen?

We aren't using the parallel codegen support from LLVM itself but rather that which we've implemented ourselves, so although it is indeed needed for truly parallel codegen it's not something we use. In https://github.com/rust-lang/llvm/commit/664319f7cca6374c320adf0ebe38305bca50ac68 the `thread` class was just stubbed out to something small.

**Comment by alexcrichton** [comment]
@bors: r=nikomatsakis f9b267a force

(gonna see if we can get the bots to not build LLVM a million times)

**Comment by alexcrichton** [comment]
:crescent_moon:

**Pull request merged**